### PR TITLE
Implement more opcodes

### PR DESCRIPTION
**Pull request by Martmists-GH**
- Implements all `BINARY_*`, `INPLACE_*`, `UNARY_*`, and `COMPARE_OP` (except for 'exception match' which requires isinstance to be implemented)
- Implements the `iter()` builtin
- Adds a `magicMethod` utility function to easily call magic methods with param if needed
- Implements `IMPORT_*` (lacking a proper implementation of IMPORT_NAME due to there not being a reliable way to get all members of an object)

### CODE REVIEW

**Comment by avayert** [comment]
Where's the two-arg form of `iter`

**Comment by Fuyukai** [comment]
I'm going to put off reviewing this for a bit until I redo attributes/methods (again!) because it'll probably have a lot of knock-on effects of the code here.

**Comment by Fuyukai** [comment]
Notably, your magicMethod helper wouldn't be needed (you would use py<Method> generally).

File: src/main/kotlin/interpreter/functions/IterBuiltinFunction.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    )
+
+    override fun callFunction(kwargs: Map<String, PyObject>): PyObject {
+        val obb = kwargs["obb"] ?: error("Built-in signature mismatch!")
+        val sentinel = kwargs["sentinel"]
+        val iter = obb.pyGetAttribute("__iter__")
+        if (iter !is PyCallable) {
+            Exceptions.TYPE_ERROR("__iter__ is not callable").throwKy()
+        }
+        return iter.runCallable(if (sentinel == null) listOf(obb) else listOf(obb, sentinel))
```
Comment (Fuyukai) [inline src/main/kotlin/interpreter/functions/IterBuiltinFunction.kt]: `__iter__` will be a bound method, so don't pass `obb`.

(Either way, it'd be wrong, as args are passed right to left not left to right so it'd be sentinel, obb).

File: src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt [hunk 2]
```diff
@@ -80,6 +97,31 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
...
+    fun magicMethod(obj: PyObject, magicName: String, param: PyObject?) {
+        val fn = obj.pyGetAttribute(magicName)
+        if (fn !is PyCallable) {
+            Exceptions.TYPE_ERROR("'${obj.type.name}'.$magicName is not callable.").throwKy()
+        }
+        val result = if (param != null) fn.runCallable(listOf(param)) else fn.runCallable(listOf())
+        stack.push(result)
+    }
+
+    fun magicMethod(obj: PyObject, magicName: String) = magicMethod(obj, magicName, null)
```
Comment (Fuyukai) [inline src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt]: You don't need this. Do `param: PyObject = null` and annotate with `@JvmOverloads`.

File: src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt [hunk 3]
```diff
@@ -132,8 +216,16 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
...
+                    // Unary operations
+                    InstructionOpcode.UNARY_POSITIVE -> unaryPostive(param)
+                    InstructionOpcode.UNARY_NEGATIVE -> unaryNegative(param)
+                    InstructionOpcode.UNARY_NOT -> unaryNot(param)
+                    InstructionOpcode.UNARY_INVERT -> unaryInvert(param)
+                    InstructionOpcode.GET_ITER -> getIter(param)
+                    InstructionOpcode.GET_YIELD_FROM_ITER -> getYieldIter(param)
 
+                    InstructionOpcode.MAKE_FUNCTION -> makeFunction(param)
+                    InstructionOpcode.COMPARE_OP -> compareOp(param)
```
Comment (Fuyukai) [inline src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt:257]: Should be on its own newline

File: src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt [hunk 4]
```diff
@@ -231,18 +323,73 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
...
     fun makeFunction(arg: Byte) {
+        // toInt to use bitwise `and`
+        val flags = arg.toInt()
         val qualifiedName = stack.pop()
         require(qualifiedName is PyString) { "Function qualified name was not string!" }
 
         val code = stack.pop()
         require(code is PyCodeObject) { "Function code was not a code object!" }
-
+        if (flags and 8 == 8) {
```
Comment (Fuyukai) [inline src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt]: Don't make this magic numbers. Make them const vals somewhere.

File: src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt [hunk 5]
```diff
@@ -231,18 +323,73 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
...
+        function.module = function.module
         stack.push(function)
         bytecodePointer += 1
     }
 
+    // Imports
+    /**
+     * IMPORT_NAME
+     */
+    fun importName(arg: Byte) {
```
Comment (Fuyukai) [inline src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt:384]: This entire thing should be a TODO(), not a fake implementation.

File: src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt [hunk 6]
```diff
@@ -231,18 +323,73 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
...
+        bytecodePointer += 1
+    }
+
+    /**
+     * IMPORT_STAR
+     */
+    fun importStar(arg: Byte) {
+        val module = stack.pop()
+
+        // Assuming internalDict is __dir__
```
Comment (Fuyukai) [inline src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt]: `pyDefaultDir()` was added in methods and attributes v3 PR, which should be what you want for now.

File: src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt [hunk 7]
```diff
@@ -329,16 +476,120 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
...
+            BinaryOp.MODULO -> "__mod__"
+            BinaryOp.SUBTRACT -> "__sub__"
+            BinaryOp.SUBSCR -> "__getitem__"
+            BinaryOp.RSHIFT -> "__rshift__"
+            BinaryOp.AND -> "__and__"
+            BinaryOp.XOR -> "__xor__"
+            BinaryOp.OR -> "__or__"
+            else -> error("This should never happen!")
+        }
+        magicMethod(o1, magic, o2, "__r" + magic.substring(2))
```
Comment (Fuyukai) [inline src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt:520]: Probably should add a comment explaining why you do `__r + whatever`

File: src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt [hunk 8]
```diff
@@ -329,16 +476,120 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
...
+            6 -> magicMethod(top, "__contains__", second)
+            7 -> {
+                magicMethod(top, "__contains__", second)
+                stack.push(if (stack.pop() == PyBool.TRUE) PyBool.FALSE else PyBool.TRUE)
             }
-            // Kept for future-proofing.
-            else -> TODO("Unsupported binary op $type")
+            8 -> stack.push(if (top.hashCode() == second.hashCode()) PyBool.TRUE else PyBool.FALSE)
+            9 -> stack.push(if (top.hashCode() != second.hashCode()) PyBool.TRUE else PyBool.FALSE)
+            10 -> TODO("exception match COMPARE_OP")
```
Comment (Fuyukai) [inline src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt]: Else branch

File: src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt [hunk 9]
```diff
@@ -329,16 +476,120 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
...
+        }
+        magicMethod(o1, magic, o2)
+        bytecodePointer += 1
+    }
+
+    fun compareOp(arg: Byte) {
+        val top = stack.pop()
+        val second = stack.pop()
+        when (arg.toInt()) {
+            0 -> magicMethod(top, "__lt__", second, "__ge__")
```
Comment (Fuyukai) [inline src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt]: Could do with turning these from magic numbers into const vals

File: src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt [hunk 10]
```diff
@@ -329,16 +476,120 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
...
-            // Kept for future-proofing.
-            else -> TODO("Unsupported binary op $type")
+            8 -> stack.push(if (top.hashCode() == second.hashCode()) PyBool.TRUE else PyBool.FALSE)
+            9 -> stack.push(if (top.hashCode() != second.hashCode()) PyBool.TRUE else PyBool.FALSE)
+            10 -> TODO("exception match COMPARE_OP")
         }
+    }
+
+    // Unary operators
+    fun getYieldIter(param: Byte) {
```
Comment (Fuyukai) [inline src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt:585]: All of these need comments.

File: src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt [hunk 11]
```diff
@@ -329,16 +476,120 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
...
+        val top = stack.pop()
+        // TODO: Generator/Coroutine check
+        if (false) {
```
Comment (Fuyukai) [inline src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt]: Please don't impl stuff like this. TODO() stub it out.

File: src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt [hunk 12]
```diff
@@ -329,16 +476,120 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
...
+        bytecodePointer += 1
+    }
+
+    fun getIter(param: Byte) {
+        val top = stack.pop()
+        magicMethod(top, "__iter__")
+        bytecodePointer += 1
+    }
+
+    fun unaryInvert(param: Byte) {
```
Comment (Fuyukai) [inline src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt]: Most of these could be turned into a single function encapsulating an enum.

**Review by Fuyukai — CHANGES_REQUESTED** [review]

File: src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt (hunk 2 above)
Comment (ilevn) [reply src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt]: Or just make PyObject nullable?

File: src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt (hunk 8 above)
Comment (Fuyukai) [reply src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt]: Else branch should be an error(), because it usually means we didn't handle a param.

**Comment by ilevn** [comment]
Spotless check is failing.

File: src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt [hunk 13]
```diff
@@ -54,6 +77,27 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
             SET,
             STRING,
         }
+
+        object FunctionFlags {
+            val POSITIONAL_DEFAULT = 1
```
Comment (Fuyukai) [inline src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt]: These should all be const vals.

**Review by Fuyukai — CHANGES_REQUESTED** [review]

**Review by Fuyukai — APPROVED** [review]

**Pull request merged**